"""
AIVA gRPC æœå‹™å¯¦ç¾ - V2 çµ±ä¸€æ¶æ§‹
========================================

åŸºæ–¼è‡ªå‹•ç”Ÿæˆçš„ Protocol Buffers å¯¦ç¾æ ¸å¿ƒ gRPC æœå‹™
- TaskService: ä»»å‹™ç®¡ç†æœå‹™
- CrossLanguageService: è·¨èªè¨€é€šä¿¡æœå‹™

åŠŸèƒ½:
- çµ±ä¸€çš„ gRPC æœå‹™ç«¯é»
- èˆ‡ MQ ç³»çµ±æ•´åˆ
- åˆ†æ•£å¼è¿½è¹¤æ”¯æ´
- è‡ªå‹•æœå‹™ç™¼ç¾
"""

import asyncio
import logging
import uuid
from datetime import datetime
from typing import AsyncIterator, Dict, Optional

import grpc
from grpc import aio  # type: ignore

# gRPC ç”Ÿæˆçš„å­˜æ ¹ (éœ€è¦å…ˆç·¨è­¯ .proto æª”æ¡ˆ)
try:
    from services.aiva_common.grpc.generated.python import aiva_pb2, aiva_pb2_grpc
except ImportError:
    # å¦‚æœé‚„æœªç·¨è­¯ï¼Œå…ˆä½¿ç”¨æ¨¡æ“¬å®šç¾©
    logging.warning("gRPC å­˜æ ¹æœªæ‰¾åˆ°ï¼Œä½¿ç”¨æ¨¡æ“¬å®šç¾©ã€‚è«‹åŸ·è¡Œ compile_protos.py")
    aiva_pb2 = None
    aiva_pb2_grpc = None

from services.aiva_common.schemas.generated.messaging import AivaMessage
from services.aiva_common.messaging.compatibility_layer import message_broker
from services.aiva_common.enums import Topic, ModuleName

logger = logging.getLogger(__name__)


class TaskServiceImplementation(aiva_pb2_grpc.TaskServiceServicer if aiva_pb2_grpc else object):
    """ä»»å‹™ç®¡ç†æœå‹™å¯¦ç¾"""
    
    def __init__(self):
        self.active_tasks = {}  # æ´»èºä»»å‹™è¿½è¹¤
        self.task_results = {}  # ä»»å‹™çµæœç·©å­˜
    
    async def CreateTask(self, request, context):
        """å‰µå»ºæ–°ä»»å‹™"""
        try:
            task_id = str(uuid.uuid4())
            
            # è½‰æ› gRPC è«‹æ±‚ç‚ºå…§éƒ¨æ ¼å¼
            task_config = {
                "task_id": task_id,
                "task_type": request.task_type,
                "target": {
                    "url": request.target.url,
                    "host": request.target.host,
                    "port": request.target.port
                },
                "parameters": dict(request.parameters),
                "priority": request.priority,
                "timeout": request.timeout
            }
            
            # é€é MQ ç™¼ä½ˆä»»å‹™
            message = message_broker.publish(
                topic=Topic.TASK_SCAN_START,
                payload=task_config,
                source_module=ModuleName.API_GATEWAY,
                target_module=ModuleName.SCAN,
                trace_id=str(uuid.uuid4())
            )
            
            # è¨˜éŒ„æ´»èºä»»å‹™
            self.active_tasks[task_id] = {
                "config": task_config,
                "created_at": datetime.now(),
                "status": "PENDING",
                "message_id": message.header.message_id,
                "trace_id": message.trace_id
            }
            
            logger.info(f"ğŸ“‹ å‰µå»ºä»»å‹™: {task_id} (é¡å‹: {request.task_type})")
            
            # å›å‚³ gRPC éŸ¿æ‡‰
            if aiva_pb2:
                return aiva_pb2.AIVAResponse(
                    request_id=task_id,
                    success=True,
                    result={"task_id": task_id, "status": "created"},
                    timestamp=datetime.now()
                )
            else:
                # æ¨¡æ“¬å›å‚³
                return {"success": True, "task_id": task_id}
                
        except Exception as e:
            logger.error(f"âŒ å‰µå»ºä»»å‹™å¤±æ•—: {e}")
            context.set_code(grpc.StatusCode.INTERNAL)
            context.set_details(f"ä»»å‹™å‰µå»ºå¤±æ•—: {e}")
            return aiva_pb2.AIVAResponse(success=False, error_message=str(e)) if aiva_pb2 else {"success": False}
    
    async def GetTaskStatus(self, request, context):
        """ç²å–ä»»å‹™ç‹€æ…‹"""
        try:
            task_id = request.request_id
            
            if task_id not in self.active_tasks:
                context.set_code(grpc.StatusCode.NOT_FOUND)
                context.set_details(f"ä»»å‹™ä¸å­˜åœ¨: {task_id}")
                return aiva_pb2.AIVAResponse(success=False, error_message="ä»»å‹™ä¸å­˜åœ¨") if aiva_pb2 else {"success": False}
            
            task_info = self.active_tasks[task_id]
            
            logger.info(f"ğŸ“Š æŸ¥è©¢ä»»å‹™ç‹€æ…‹: {task_id} -> {task_info['status']}")
            
            if aiva_pb2:
                return aiva_pb2.TaskResult(
                    task_id=task_id,
                    status=getattr(aiva_pb2.TaskStatus, task_info['status'], aiva_pb2.TaskStatus.TASK_STATUS_PENDING),
                    started_at=task_info['created_at'],
                    metadata={"trace_id": task_info['trace_id']}
                )
            else:
                return {"task_id": task_id, "status": task_info['status']}
                
        except Exception as e:
            logger.error(f"âŒ æŸ¥è©¢ä»»å‹™ç‹€æ…‹å¤±æ•—: {e}")
            context.set_code(grpc.StatusCode.INTERNAL)
            context.set_details(f"ç‹€æ…‹æŸ¥è©¢å¤±æ•—: {e}")
            return {"success": False}
    
    async def CancelTask(self, request, context):
        """å–æ¶ˆä»»å‹™"""
        try:
            task_id = request.request_id
            
            if task_id not in self.active_tasks:
                context.set_code(grpc.StatusCode.NOT_FOUND)
                context.set_details(f"ä»»å‹™ä¸å­˜åœ¨: {task_id}")
                return aiva_pb2.AIVAResponse(success=False, error_message="ä»»å‹™ä¸å­˜åœ¨") if aiva_pb2 else {"success": False}
            
            # é€é MQ ç™¼é€å–æ¶ˆå‘½ä»¤
            message_broker.publish(
                topic=Topic.COMMAND_TASK_CANCEL,
                payload={"task_id": task_id, "reason": "user_cancelled"},
                source_module=ModuleName.API_GATEWAY,
                target_module=ModuleName.CORE,
                trace_id=self.active_tasks[task_id]["trace_id"]
            )
            
            # æ›´æ–°ä»»å‹™ç‹€æ…‹
            self.active_tasks[task_id]["status"] = "CANCELLED"
            
            logger.info(f"âŒ å–æ¶ˆä»»å‹™: {task_id}")
            
            if aiva_pb2:
                return aiva_pb2.AIVAResponse(
                    request_id=task_id,
                    success=True,
                    result={"task_id": task_id, "status": "cancelled"},
                    timestamp=datetime.now()
                )
            else:
                return {"success": True, "task_id": task_id}
                
        except Exception as e:
            logger.error(f"âŒ å–æ¶ˆä»»å‹™å¤±æ•—: {e}")
            context.set_code(grpc.StatusCode.INTERNAL)
            context.set_details(f"ä»»å‹™å–æ¶ˆå¤±æ•—: {e}")
            return {"success": False}
    
    async def StreamTaskProgress(self, request, context):
        """ä¸²æµä»»å‹™é€²åº¦"""
        try:
            task_id = request.request_id
            
            if task_id not in self.active_tasks:
                context.set_code(grpc.StatusCode.NOT_FOUND)
                context.set_details(f"ä»»å‹™ä¸å­˜åœ¨: {task_id}")
                return
            
            logger.info(f"ğŸ“¡ é–‹å§‹ä¸²æµä»»å‹™é€²åº¦: {task_id}")
            
            # æ¨¡æ“¬é€²åº¦æ›´æ–°ï¼ˆå¯¦éš›æ‡‰è©²å¾ MQ æ¥æ”¶ï¼‰
            progress_steps = [
                ("RUNNING", "ä»»å‹™é–‹å§‹åŸ·è¡Œ"),
                ("PROGRESS", "æƒæä¸­... 25%"),
                ("PROGRESS", "æƒæä¸­... 50%"),
                ("PROGRESS", "æƒæä¸­... 75%"),
                ("COMPLETED", "ä»»å‹™å®Œæˆ")
            ]
            
            for status, message in progress_steps:
                if context.cancelled():
                    break
                
                # æ›´æ–°ä»»å‹™ç‹€æ…‹
                self.active_tasks[task_id]["status"] = status
                
                if aiva_pb2:
                    response = aiva_pb2.AIVAResponse(
                        request_id=task_id,
                        success=True,
                        result={"status": status, "message": message},
                        timestamp=datetime.now()
                    )
                    yield response
                else:
                    yield {"task_id": task_id, "status": status, "message": message}
                
                # æ¨¡æ“¬è™•ç†æ™‚é–“
                await asyncio.sleep(2)
            
            logger.info(f"âœ… ä»»å‹™é€²åº¦ä¸²æµå®Œæˆ: {task_id}")
            
        except Exception as e:
            logger.error(f"âŒ ä»»å‹™é€²åº¦ä¸²æµå¤±æ•—: {e}")
            context.set_code(grpc.StatusCode.INTERNAL)
            context.set_details(f"é€²åº¦ä¸²æµå¤±æ•—: {e}")


class CrossLanguageServiceImplementation(aiva_pb2_grpc.CrossLanguageServiceServicer if aiva_pb2_grpc else object):
    """è·¨èªè¨€é€šä¿¡æœå‹™å¯¦ç¾"""
    
    def __init__(self):
        self.service_registry = {}  # æœå‹™è¨»å†Šè¡¨
        self.active_connections = {}  # æ´»èºé€£ç·š
    
    async def ExecuteTask(self, request, context):
        """åŸ·è¡Œè·¨èªè¨€ä»»å‹™"""
        try:
            task_type = request.task
            trace_id = request.trace_id or str(uuid.uuid4())
            
            # æ ¹æ“šä»»å‹™é¡å‹è·¯ç”±åˆ°å°æ‡‰èªè¨€æœå‹™
            if "scan" in task_type.lower():
                target_module = ModuleName.SCAN
                topic = Topic.TASK_SCAN_START
            elif "function" in task_type.lower():
                target_module = ModuleName.FUNCTION
                topic = Topic.TASK_FUNCTION_START
            else:
                target_module = ModuleName.CORE
                topic = Topic.TASK_SCAN_START
            
            # é€éçµ±ä¸€ MQ ç³»çµ±åŸ·è¡Œ
            message = message_broker.publish(
                topic=topic,
                payload=dict(request.parameters),
                source_module=ModuleName.API_GATEWAY,
                target_module=target_module,
                trace_id=trace_id
            )
            
            logger.info(f"ğŸ”„ åŸ·è¡Œè·¨èªè¨€ä»»å‹™: {task_type} -> {target_module}")
            
            if aiva_pb2:
                return aiva_pb2.AIVAResponse(
                    request_id=request.request_id,
                    success=True,
                    result={"task_type": task_type, "message_id": message.header.message_id},
                    timestamp=datetime.now()
                )
            else:
                return {"success": True, "task_type": task_type}
                
        except Exception as e:
            logger.error(f"âŒ è·¨èªè¨€ä»»å‹™åŸ·è¡Œå¤±æ•—: {e}")
            context.set_code(grpc.StatusCode.INTERNAL)
            context.set_details(f"ä»»å‹™åŸ·è¡Œå¤±æ•—: {e}")
            return {"success": False}
    
    async def HealthCheck(self, request, context):
        """å¥åº·æª¢æŸ¥"""
        try:
            logger.info("ğŸ’“ gRPC å¥åº·æª¢æŸ¥")
            
            # æª¢æŸ¥ MQ é€£æ¥ç‹€æ…‹
            mq_status = "healthy"  # å¯¦éš›æ‡‰è©²æª¢æŸ¥ MQ é€£æ¥
            
            health_info = {
                "status": "healthy",
                "timestamp": datetime.now().isoformat(),
                "mq_status": mq_status,
                "active_tasks": len(getattr(self, 'task_service', {}).get('active_tasks', {})),
                "service_version": "1.1.0"
            }
            
            if aiva_pb2:
                return aiva_pb2.AIVAResponse(
                    request_id=request.request_id,
                    success=True,
                    result=health_info,
                    timestamp=datetime.now()
                )
            else:
                return {"success": True, "health": health_info}
                
        except Exception as e:
            logger.error(f"âŒ å¥åº·æª¢æŸ¥å¤±æ•—: {e}")
            return {"success": False, "error": str(e)}
    
    async def GetServiceInfo(self, request, context):
        """ç²å–æœå‹™è³‡è¨Š"""
        try:
            service_info = {
                "service_name": "AIVA gRPC Server",
                "version": "1.1.0",
                "supported_languages": ["Python", "Go", "TypeScript", "Rust"],
                "supported_protocols": ["gRPC", "MQ"],
                "features": [
                    "çµ±ä¸€ä»»å‹™ç®¡ç†",
                    "è·¨èªè¨€é€šä¿¡",
                    "åˆ†æ•£å¼è¿½è¹¤",
                    "MQ æ•´åˆ",
                    "é›™å‘ä¸²æµ"
                ],
                "endpoints": [
                    "TaskService/CreateTask",
                    "TaskService/GetTaskStatus", 
                    "TaskService/CancelTask",
                    "TaskService/StreamTaskProgress",
                    "CrossLanguageService/ExecuteTask",
                    "CrossLanguageService/HealthCheck",
                    "CrossLanguageService/GetServiceInfo",
                    "CrossLanguageService/BidirectionalStream"
                ]
            }
            
            logger.info("â„¹ï¸ è¿”å›æœå‹™è³‡è¨Š")
            
            if aiva_pb2:
                return aiva_pb2.AIVAResponse(
                    request_id=request.request_id,
                    success=True,
                    result=service_info,
                    timestamp=datetime.now()
                )
            else:
                return {"success": True, "service_info": service_info}
                
        except Exception as e:
            logger.error(f"âŒ ç²å–æœå‹™è³‡è¨Šå¤±æ•—: {e}")
            return {"success": False}
    
    async def BidirectionalStream(self, request_iterator, context):
        """é›™å‘ä¸²æµé€šä¿¡"""
        try:
            logger.info("ğŸ”„ é–‹å§‹é›™å‘ä¸²æµé€šä¿¡")
            
            async for request in request_iterator:
                try:
                    # è™•ç†æ¥æ”¶åˆ°çš„è¨Šæ¯
                    task = request.task
                    trace_id = request.trace_id or str(uuid.uuid4())
                    
                    logger.info(f"ğŸ“¥ æ¥æ”¶ä¸²æµè¨Šæ¯: {task}")
                    
                    # é€é MQ è™•ç†ä¸¦ç²å–éŸ¿æ‡‰
                    message = message_broker.publish(
                        topic=Topic.TASK_SCAN_START,
                        payload=dict(request.parameters),
                        source_module=ModuleName.API_GATEWAY,
                        target_module=ModuleName.SCAN,
                        trace_id=trace_id
                    )
                    
                    # å›å‚³è™•ç†çµæœ
                    if aiva_pb2:
                        response = aiva_pb2.AIVAResponse(
                            request_id=request.request_id,
                            success=True,
                            result={"processed_task": task, "message_id": message.header.message_id},
                            timestamp=datetime.now()
                        )
                        yield response
                    else:
                        yield {"success": True, "processed_task": task}
                    
                except Exception as e:
                    logger.error(f"âŒ è™•ç†ä¸²æµè¨Šæ¯å¤±æ•—: {e}")
                    if aiva_pb2:
                        yield aiva_pb2.AIVAResponse(success=False, error_message=str(e))
                    else:
                        yield {"success": False, "error": str(e)}
            
            logger.info("âœ… é›™å‘ä¸²æµé€šä¿¡çµæŸ")
            
        except Exception as e:
            logger.error(f"âŒ é›™å‘ä¸²æµé€šä¿¡å¤±æ•—: {e}")


class AIVAGRPCServer:
    """AIVA gRPC æœå‹™å™¨"""
    
    def __init__(self, host: str = "0.0.0.0", port: int = 50051):
        self.host = host
        self.port = port
        self.server = None
        self.task_service = TaskServiceImplementation()
        self.cross_language_service = CrossLanguageServiceImplementation()
        
        # å…±äº«æœå‹™å¼•ç”¨
        self.cross_language_service.task_service = self.task_service
    
    async def start(self):
        """å•Ÿå‹• gRPC æœå‹™å™¨"""
        try:
            self.server = aio.server()
            
            if aiva_pb2_grpc:
                # è¨»å†Šæœå‹™
                aiva_pb2_grpc.add_TaskServiceServicer_to_server(
                    self.task_service, self.server
                )
                aiva_pb2_grpc.add_CrossLanguageServiceServicer_to_server(
                    self.cross_language_service, self.server
                )
            
            # æ·»åŠ ç›£è½ç«¯å£
            listen_addr = f"{self.host}:{self.port}"
            self.server.add_insecure_port(listen_addr)
            
            # å•Ÿå‹•æœå‹™å™¨
            await self.server.start()
            logger.info(f"ğŸš€ AIVA gRPC æœå‹™å™¨å•Ÿå‹•: {listen_addr}")
            logger.info("ğŸ“‹ æ”¯æ´æœå‹™: TaskService, CrossLanguageService")
            
            # ç­‰å¾…çµ‚æ­¢
            await self.server.wait_for_termination()
            
        except Exception as e:
            logger.error(f"âŒ gRPC æœå‹™å™¨å•Ÿå‹•å¤±æ•—: {e}")
            raise
    
    async def stop(self):
        """åœæ­¢ gRPC æœå‹™å™¨"""
        if self.server:
            await self.server.stop(5)
            logger.info("ğŸ›‘ AIVA gRPC æœå‹™å™¨å·²åœæ­¢")


async def main():
    """ä¸»ç¨‹å¼å…¥å£"""
    logging.basicConfig(level=logging.INFO)
    
    server = AIVAGRPCServer()
    
    try:
        await server.start()
    except KeyboardInterrupt:
        logger.info("ğŸ”´ æ¥æ”¶åˆ°ä¸­æ–·ä¿¡è™Ÿ")
        await server.stop()


if __name__ == "__main__":
    asyncio.run(main())